### PR TITLE
EDM-3720: EDM-3725,EDM-3729,EDM-3733

### PR DESCRIPTION
**Pull request by asafbennatan**
adds vulnerabilities e2e

### CODE REVIEW

File: test/e2e/vulnerability/vulnerability_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+	}
+
+	if trustifySvc != nil {
+		logrus.Info("🔄 [AfterSuite] Stopping Trustify service")
+		err := trustifySvc.Stop(ctx)
+		if err != nil {
+			logrus.Warnf("Failed to stop Trustify: %v", err)
+		}
+	}
+})
```
[the remark this answers is not in the document]
Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_suite_test.go:108]: we do not handle ginkgo -p ATM

File: test/e2e/vulnerability/README.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+The vulnerability feature provides:
+- Organization-level vulnerability summary and list endpoints
+- Device-level vulnerability queries
+- Fleet-level vulnerability aggregation
+- CVE impact analysis (blast radius)
+- CLI commands for vulnerability management
+
+## Prerequisites
+
+1. **Running FlightCtl cluster** (KIND or OpenShift)
```
Comment (amalykhi) [inline test/e2e/vulnerability/README.md]: ACM/ quadlets also?

File: test/e2e/vulnerability/vulnerability_suite_test.go [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+	Expect(err).ToNot(HaveOccurred())
+
+	harness.SetTestContext(suiteCtx)
+
+	GinkgoWriter.Printf("✅ [AfterEach] Worker %d: Test cleanup completed\n", workerID)
+})
+
+// restoreAPIConfiguration restores the API and periodic services to their original state.
+// This is called in AfterSuite to clean up after the tests.
+func restoreAPIConfiguration(ctx context.Context) error {
```
Comment (amalykhi) [inline test/e2e/vulnerability/vulnerability_suite_test.go:140]: Are all this functions test/Suite specific? or can be moved to the helper/harness folder?

File: test/e2e/vulnerability/vulnerability_test.go [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+			Expect(err).ToNot(HaveOccurred())
+
+			harness.WaitForVulnerabilitySyncComplete(v1alpha1Client, syncWaitTimeoutExtended, syncInterval)
+		})
+
+		AfterEach(func() {
+			_ = harness.DeleteDeviceIgnoreNotFound(deviceName)
+		})
+
+		It("should support sorting by severity", func() {
```
Comment (amalykhi) [inline test/e2e/vulnerability/vulnerability_test.go]: Could these Its be combined to one/less "It" and steps separated by "By"? 
Every It require separate Polarion Case

File: test/e2e/vulnerability/vulnerability_test.go [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+			Expect(err).ToNot(HaveOccurred())
+
+			harness.WaitForVulnerabilitySyncComplete(v1alpha1Client, syncWaitTimeoutExtended, syncInterval)
+		})
+
+		AfterEach(func() {
+			_ = harness.DeleteDeviceIgnoreNotFound(deviceName)
+		})
+
+		It("should return impact data for Critical CVE (CVE-2021-35942 glibc)", func() {
```
Comment (amalykhi) [inline test/e2e/vulnerability/vulnerability_test.go]: Could these Its be combined to one/less "It" and steps separated by "By"?

File: test/e2e/vulnerability/vulnerability_test.go [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+			resp, err := v1alpha1Client.GetVulnerabilityImpactWithResponse(harness.Context, "CVE-9999-99999", nil)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(resp.StatusCode()).To(Equal(200))
+			Expect(resp.JSON200).ToNot(BeNil())
+			Expect(resp.JSON200.Items).To(BeEmpty(), "Unknown CVE should have empty blast radius")
+		})
+	})
+
+	Describe("Negative and Edge Cases", func() {
+		It("should return 404 for non-existent device", func() {
```
Comment (amalykhi) [inline test/e2e/vulnerability/vulnerability_test.go]: Could these Its be combined to one/less "It" and steps separated by "By"?

File: test/e2e/vulnerability/vulnerability_test.go [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+			Expect(err).ToNot(HaveOccurred())
+
+			harness.WaitForVulnerabilitySyncComplete(v1alpha1Client, syncWaitTimeoutExtended, syncInterval)
+		})
+
+		AfterEach(func() {
+			_ = harness.DeleteDeviceIgnoreNotFound(deviceName)
+		})
+
+		It("should list org vulnerabilities with 'flightctl get vulnerability'", func() {
```
Comment (amalykhi) [inline test/e2e/vulnerability/vulnerability_test.go]: Could these Its be combined to one/less "It" and steps separated by "By"?

File: test/e2e/vulnerability_gating/vulnerability_gating_test.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+		var err error
+		v1alpha1Client, err = v1alpha1client.NewClientWithResponses(
+			harness.Server+"/api/v1alpha1",
+			v1alpha1client.WithHTTPClient(harness.HTTPClient),
+		)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	Describe("when Trustify is not configured", func() {
+		It("should return 501 for GetVulnerabilitySummary", func() {
```
Comment (amalykhi) [inline test/e2e/vulnerability_gating/vulnerability_gating_test.go]: Could these Its be combined to one/less "It" and steps separated by "By"?

File: test/harness/e2e/harness_vulnerability.go [hunk 9]
```diff
@@ -0,0 +1,603 @@
...
+// truncateDigest returns a truncated version of a digest for logging
+func truncateDigest(digest string) string {
+	if len(digest) > 25 {
+		return digest[:25] + "..."
+	}
+	return digest
+}
+
+// CreateFleetWithSelector creates a fleet with the specified label selector.
+func (h *Harness) CreateFleetWithSelector(fleetName string, labelSelector map[string]string) error {
```
Comment (amalykhi) [inline test/harness/e2e/harness_vulnerability.go]: not specific to vulnerability. Could be moved to harness_fleet.go

File: test/harness/e2e/harness_vulnerability.go [hunk 10]
```diff
@@ -0,0 +1,603 @@
...
+	if resp.StatusCode() != 201 {
+		return fmt.Errorf("unexpected status creating fleet %s: %d", fleetName, resp.StatusCode())
+	}
+
+	logrus.Infof("Created fleet %s with selector %v", fleetName, labelSelector)
+	return nil
+}
+
+// DeleteDeviceIgnoreNotFound deletes a device, ignoring 404 errors.
+func (h *Harness) DeleteDeviceIgnoreNotFound(deviceName string) error {
```
Comment (amalykhi) [inline test/harness/e2e/harness_vulnerability.go]: not specific to vulnerability. Could be moved to harness_device.go or resource_factory.go

File: test/harness/e2e/harness_vulnerability.go [hunk 11]
```diff
@@ -0,0 +1,603 @@
...
+		return fmt.Errorf("failed to delete device %s: %w", deviceName, err)
+	}
+	if resp.StatusCode() != 200 && resp.StatusCode() != 404 {
+		return fmt.Errorf("unexpected status deleting device %s: %d", deviceName, resp.StatusCode())
+	}
+	return nil
+}
+
+// DeleteFleetIgnoreNotFound deletes a fleet, ignoring 404 errors.
+func (h *Harness) DeleteFleetIgnoreNotFound(fleetName string) error {
```
Comment (amalykhi) [inline test/harness/e2e/harness_vulnerability.go]: not specific to vulnerability. Could be moved to harness_fleet.go or resource_factory.go

File: test/harness/e2e/harness_vulnerability.go [hunk 12]
```diff
@@ -0,0 +1,603 @@
...
+		return nil, err
+	}
+	if resp.JSON200 == nil {
+		return nil, fmt.Errorf("unexpected response: %d", resp.StatusCode())
+	}
+	return resp.JSON200, nil
+}
+
+// GetDeviceEvents retrieves events for a device, optionally filtered by reason.
+func (h *Harness) GetDeviceEvents(deviceName string, reason *v1beta1.EventReason) ([]v1beta1.Event, error) {
```
Comment (amalykhi) [inline test/harness/e2e/harness_vulnerability.go]: not specific to vulnerability. Currently we have only RunGetEvents() in harness.go
May be move to it or add another harness_events.go

File: test/harness/e2e/harness_vulnerability.go [hunk 13]
```diff
@@ -0,0 +1,603 @@
...
+		return nil, err
+	}
+	if resp.JSON200 == nil {
+		return nil, fmt.Errorf("unexpected response: %d", resp.StatusCode())
+	}
+	return resp.JSON200.Items, nil
+}
+
+// HasEventWithReason checks if the event list contains an event with the specified reason.
+func HasEventWithReason(events []v1beta1.Event, reason v1beta1.EventReason) bool {
```
Comment (amalykhi) [inline test/harness/e2e/harness_vulnerability.go]: not specific to vulnerability. Currently we have only RunGetEvents() in harness.go
May be move to it or add another harness_events.go

File: test/harness/e2e/harness_vulnerability.go [hunk 14]
```diff
@@ -0,0 +1,603 @@
...
+	return &v
+}
+
+// DeviceVulnerabilitySortOrderPtr converts a string to a GetDeviceVulnerabilitiesParamsOrder pointer.
+func DeviceVulnerabilitySortOrderPtr(s string) *v1alpha1.GetDeviceVulnerabilitiesParamsOrder {
+	v := v1alpha1.GetDeviceVulnerabilitiesParamsOrder(s)
+	return &v
+}
+
+func (h *Harness) dumpDevicesList() {
```
Comment (amalykhi) [inline test/harness/e2e/harness_vulnerability.go:373]: it is just for logging and visibility, right?

Comment (asafbennatan) [reply test/harness/e2e/harness_vulnerability.go:373]: yes

File: test/harness/e2e/harness_vulnerability.go (hunk 13 above)
Comment (asafbennatan) [reply test/harness/e2e/harness_vulnerability.go]: done

File: test/harness/e2e/harness_vulnerability.go (hunk 12 above)
Comment (asafbennatan) [reply test/harness/e2e/harness_vulnerability.go]: done

File: test/harness/e2e/harness_vulnerability.go (hunk 11 above)
Comment (asafbennatan) [reply test/harness/e2e/harness_vulnerability.go]: done

File: test/harness/e2e/harness_vulnerability.go (hunk 10 above)
Comment (asafbennatan) [reply test/harness/e2e/harness_vulnerability.go]: done

File: test/harness/e2e/harness_vulnerability.go (hunk 9 above)
Comment (asafbennatan) [reply test/harness/e2e/harness_vulnerability.go]: done

File: test/e2e/vulnerability/README.md (hunk 2 above)
Comment (asafbennatan) [reply test/e2e/vulnerability/README.md]: done

File: test/e2e/vulnerability_gating/vulnerability_gating_test.go (hunk 8 above)
Comment (asafbennatan) [reply test/e2e/vulnerability_gating/vulnerability_gating_test.go]: done

File: test/e2e/vulnerability/vulnerability_test.go (hunk 7 above)
Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_test.go]: done

File: test/e2e/vulnerability/vulnerability_test.go (hunk 6 above)
Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_test.go]: done

File: test/e2e/vulnerability/vulnerability_test.go (hunk 5 above)
Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_test.go]: done

File: test/e2e/vulnerability/vulnerability_test.go (hunk 4 above)
Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_test.go]: done

File: test/e2e/vulnerability/vulnerability_suite_test.go (hunk 3 above)
Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_suite_test.go:140]: adjusted , remaining functions are suite specific

File: test/e2e/vulnerability/vulnerability_test.go [hunk 15]
```diff
@@ -149,42 +150,40 @@ var _ = Describe("Vulnerability API", Label("vulnerability"), func() {
 			_ = harness.DeleteDeviceIgnoreNotFound(deviceName)
 		})
 
-		It("should support sorting by severity", func() {
+		It("should support sorting, pagination, and device-level queries", func() {
```
Comment (amalykhi) [inline test/e2e/vulnerability/vulnerability_test.go:153]: For all "It"s - which of them deserve "sanity" label from your point of view?
I'll convert to manual and add Polarion once approved and merged

Comment (asafbennatan) [reply test/e2e/vulnerability/vulnerability_test.go:153]: none

**Comment by amalykhi** [comment]
image

**Comment by amalykhi** [comment]
Screenshot From 2026-05-13 18-59-32
THe test run succeeded in OCP

**Review by amalykhi — APPROVED** [review]
